8_bit_pic.py

```diff
@@ -11,102 +37,352 @@
+    def __init__(self, k: str) -> None:
+        super().__init__("MOVLW", 0x3000, k, 0xff)
+
+
+TEST_PROG = """
 STATUS  equ 03h
 TRISA   equ 85h
 PORTA   equ 05h
+
+        nop
         bsf STATUS,5
         movlw 00h
         movwf TRISA
         bcf STATUS,5
-Start   movlw 02h
+
+Start:  movlw 02h
         movwf PORTA
         movlw 00h
         movwf PORTA
-        goto Star
+        goto Start
 """
 
 
-class Assembler:
+class PICAssembler(vm_types.GenericAssembler):
+
+    PSEUDOOP_RE = re.compile(r"(?P<name>[A-Za-z]\w*)\s+(?P<pseudoop>\w+)\s+(?P<operands>\S+)(\s+?P<comment>;.*)?")
+    OPCODE_RE = re.compile(r"((?P<label>[A-Za-z]\w*):\s+)?(?P<opcode>\w+)(\s+(?P<operands>\S+))?(\s+?P<comment>;.*)?")
+
     def __init__(
         self,
         program_text,
         instruction_codes: type[vm_types.GenericInstructionSet],
         word_size,
     ):
         self.text = program_text
+        self.codes = instruction_codes
+        self.word_size = word_size
+        self.word_size_bytes = word_size // vm_types.BITS_IN_BYTE
+        self.symbol_table = {"map": {}, "symbols": {}}
         self.byte_code = bytearray()
 
     def load_program(self, program_text):
         self.text = program_text
 
+    def _process_pseudoop(self, name: str, pseudoop: str, operands: list[str], line_num: int) -> None:
+        match pseudoop:
+            case "EQU":
+                # Defines symbol value.
+                if name in self.symbol_table["symbols"]:
+                    raise Exception(f"Symbol '{pseudoop}' redefinition on line {line_num}.")
+                self.symbol_table["symbols"][name] = operands[0]
+            case "SET":
+                # Defines or re-defines symbol value.
+                self.symbol_table["symbols"][name] = operands[0]
+            case "ERROR":
+                # Generates a user-defined error.
+                raise Exception(f"Error {operands[0]}.")
+            case _:
+                raise Exception(f"Unexpected pseudoop '{pseudoop}' on line {line_num}.")
+
+    def _process_opcode(self, label: str, opcode: str, operands: list[str], line_num: int, tokens: list[vm_types.AssemblerToken]) -> None:
+        if label:
+            tokens.append(LabelToken(label))
+        match opcode:
+            case "MOVWF":
+                tokens.append(MOVWFToken(operands[0]))
+            case "NOP":
+                tokens.append(NOPToken())
+            case "BCF":
+                tokens.append(BCFToken(operands[0], operands[1]))
+            case "BSF":
+                tokens.append(BSFToken(operands[0], operands[1]))
+            case "GOTO":
+                tokens.append(GOTOToken(operands[0]))
+            case "MOVLW":
+                tokens.append(MOVLWToken(operands[0]))
+            case _:
+                raise Exception(f"Unexpected opcode '{opcode}' on line {line_num}.")
+
     def tokenize(self) -> Sequence[vm_types.AssemblerToken]:
-        return []
+        tokens: list[vm_types.AssemblerToken] = []
```

Comment:
The stream of tokens idea is nice.

I guess this method should be a generator?